### PR TITLE
[5.0] Be conservative with classifying some casts that could change in the future.

### PR DESCRIPTION
**Pull request by jckarter**
- Existentials may eventually be able to conform to protocols
- Functions may grow new arms we haven't designed and implemented yet,
  which could influence how they cast in the future.

This does not directly affect the runtime ABI, but is a critical fix, since it prevents optimized code from making incorrect assumptions about how future runtimes may behave.

rdar://problem/38684452

### CODE REVIEW

**Comment by jckarter** [comment]
@swift-ci Please test

**Review by AnnaZaks — APPROVED** [review]
Does not affect ABI and is not critical for convergence, but we'd like to get it into Swift 5. Approved but will need to wait until we converge before we merge it.

**Comment by jckarter** [comment]
The OS X test failures in lldb don't look related; @fredriss do these unexpected passes ring a bell?

```
16:51:11 =============
16:51:11 Issue Details
16:51:11 =============
16:51:11 UNEXPECTED SUCCESS: test_cross_module_extension_dsym (lang/swift/unit-tests/TestSwiftUnitTests.py)
16:51:11 UNEXPECTED SUCCESS: test_dwarf (lang/swift/foundation_value_types/notification/TestSwiftFoundationTypeNotification.py)
16:51:11 UNEXPECTED SUCCESS: test_repl_in_c (lang/swift/repl_in_c/TestSwiftReplInC.py)
16:51:11 TIMEOUT: test_expressions_in_class_functions_dwarf 
```

**Comment by jckarter** [comment]
@swift-ci Please test

**Pull request merged**
